### PR TITLE
NET-6294 - v1 Agentless proxycfg datasource errors after v2 changes

### PR DESCRIPTION
**Pull request by jmurret**
### Description

This PR is intended to fix the following message when running consul in v1 kubernetes and registering a service in kubernetes which spawns a s sidecar:
```
2023-10-24T16:22:00.507Z [ERROR] agent.proxycfg: Failed to handle update from watch: kind=connect-proxy proxy=default/default/forward1-55d89888cd-5758v-forward1-sidecar-proxy service_id=default/default/forward1-55d89888cd-5758v-forward1-sidecar-proxy id=service-http-checks:default/default/forward1-55d89888cd-5758v-forward1 error="error filling agent cache: Internal cache failure: service 'forward1-55d89888cd-5758v-forward1' not in agent state"
```
The following two issues that resulted to the fix of the above were also fixed:
- `TestEnvoy/case-api-gateway-http-hostnames` was failing
- `TestHTTPRouteRetryAndTimeout` was previously running TestHTTPRouteRetryAndTimeout has a panic when it is not set. When it is set, the panic goes away.  NumRetries does not panic if it is unset, which is confusing.
```
dc1-consul-server-0-w6tcosn0t ~~ goroutine 435 [running]:
dc1-consul-server-0-w6tcosn0t ~~ github.com/hashicorp/consul/proto/private/pbconfigentry.RetryFilterFromStructs(...)
dc1-consul-server-0-w6tcosn0t ~~ 	/Users/jmurret/Documents/github/hashicorp/consul/proto/private/pbconfigentry/config_entry.gen.go:1837
dc1-consul-server-0-w6tcosn0t ~~ github.com/hashicorp/consul/proto/private/pbconfigentry.HTTPFiltersFromStructs(0x4001778a80, 0x40013ff200)
dc1-consul-server-0-w6tcosn0t ~~ 	/Users/jmurret/Documents/github/hashicorp/consul/proto/private/pbconfigentry/config_entry.gen.go:437 +0x274
dc1-consul-server-0-w6tcosn0t ~~ github.com/hashicorp/consul/proto/private/pbconfigentry.HTTPRouteRuleFromStructs(0x4001778a80, 0x4000956620)
dc1-consul-server-0-w6tcosn0t ~~ 	/Users/jmurret/Documents/github/hashicorp/consul/proto/private/pbconfigentry/config_entry.gen.go:700 +0x4c
dc1-consul-server-0-w6tcosn0t ~~ github.com/hashicorp/consul/proto/private/pbconfigentry.HTTPRouteFromStructs(0x4000ea7540, 0x400190bd00)
dc1-consul-server-0-w6tcosn0t ~~ 	/Users/jmurret/Documents/github/hashicorp/consul/proto/private/pbconfigentry/config_entry.gen.go:654 +0x27c
dc1-consul-server-0-w6tcosn0t ~~ github.com/hashicorp/consul/proto/private/pbconfigentry.ConfigEntryFromStructs({0x47fc0c0, 0x4000ea7540})
dc1-consul-server-0-w6tcosn0t ~~ 	/Users/jmurret/Documents/github/hashicorp/consul/proto/private/pbconfigentry/config_entry.go:201 +0x620
dc1-consul-server-0-w6tcosn0t ~~ github.com/hashicorp/consul/agent/consul/state.EventPayloadConfigEntry.ToSubscriptionEvent(...)
dc1-consul-server-0-w6tcosn0t ~~ 	/Users/jmurret/Documents/github/hashicorp/consul/agent/consul/state/config_entry_events.go:71
dc1-consul-server-0-w6tcosn0t ~~ github.com/hashicorp/consul/agent/submatview.(*LocalMaterializer).subscribeOnce(0x40008e5500, {0x47ec2d8?, 0x400137ee60?}, 0x40012fe160)
dc1-consul-server-0-w6tcosn0t ~~ 	/Users/jmurret/Documents/github/hashicorp/consul/agent/submatview/local_materializer.go:132 +0x304
dc1-consul-server-0-w6tcosn0t ~~ github.com/hashicorp/consul/agent/submatview.(*LocalMaterializer).Run(0x40008e5500, {0x47ec2d8, 0x400137ee60})
dc1-consul-server-0-w6tcosn0t ~~ 	/Users/jmurret/Documents/github/hashicorp/consul/agent/submatview/local_materializer.go:68 +0x60
dc1-consul-server-0-w6tcosn0t ~~ github.com/hashicorp/consul/agent/submatview.(*Store).readEntry.func1()
dc1-consul-server-0-w6tcosn0t ~~ 	/Users/jmurret/Documents/github/hashicorp/consul/agent/submatview/store.go:236 +0x50
dc1-consul-server-0-w6tcosn0t ~~ created by github.com/hashicorp/consul/agent/submatview.(*Store).readEntry in goroutine 423
dc1-consul-server-0-w6tcosn0t ~~ 	/Users/jmurret/Documents/github/hashicorp/consul/agent/submatview/store.go:235 +0x430
```
### PR Checklist

* [ ] updated test coverage
* [ ] external facing docs updated
* [ ] appropriate backport labels added
* [ ] not a security concern

### CODE REVIEW

**Review by sarahalsmiller — COMMENTED** [review]
The gateway test issues being introduced with this configuration update still doesn't really make sense to me, but your fixes look good to me on the API Gateway side of things. I'll make sure to follow up and dig more into why that code started panicing to begin with. I'll let someone with more domain knowledge look at the config updates.

**Comment by jmurret** [comment]
Verified original error message in logs is still fixed with latest version.

**Comment by jmurret** [comment]
CI passes in Enterprise as well.

File: api/config_entry_routes.go [hunk 1]
```diff
@@ -225,7 +225,7 @@ type RetryFilter struct {
 	NumRetries            *uint32
 	RetryOn               []string
 	RetryOnStatusCodes    []uint32
-	RetryOnConnectFailure *bool
```
Comment (hashi-derek) [inline api/config_entry_routes.go:228]: Why did we remove the pointer?

**Review by ishustava — APPROVED** [review]

File: api/config_entry_routes.go (hunk 1 above)
Comment (jmurret) [reply api/config_entry_routes.go:228]: `TestHTTPRouteRetryAndTimeout` failed with a panic.

File: agent/agent.go [hunk 2]
```diff
@@ -921,29 +921,13 @@ func (a *Agent) Failed() <-chan struct{} {
 	return a.apiServers.failed
 }
 
-// getProxyWatcher returns the proper implementation of the ProxyWatcher interface.
-// It will return a ProxyTracker if "resource-apis" experiment is active.  Otherwise,
-// it will return a ConfigSource.
-func (a *Agent) getProxyWatcher() xds.ProxyWatcher {
-	if a.baseDeps.UseV2Resources() {
-		a.logger.Trace("returning proxyTracker for getProxyWatcher")
```
Comment (rboyer) [inline agent/agent.go:929]: Did you want to retain this trace log in the refactored version?

File: proto/private/pbconfigentry/config_entry.gen.go [hunk 3]
```diff
@@ -1834,7 +1834,7 @@ func RetryFilterFromStructs(t *structs.RetryFilter, s *RetryFilter) {
 	s.NumRetries = *t.NumRetries
 	s.RetryOn = t.RetryOn
 	s.RetryOnStatusCodes = t.RetryOnStatusCodes
-	s.RetryOnConnectFailure = *t.RetryOnConnectFailure
```
Comment (rboyer) [inline proto/private/pbconfigentry/config_entry.gen.go:1837]: Was the desire for this `RetryFilter.RetryOnConnectFailure` to originally have been 3-state? (enabled, disabled, default)

The protobuf form in `pbconfigentry` is flagged as `bool` and the structs internal form is `*bool` prior to John's change.

`mog` is generating a `bool <-> *bool` conversion, just like it is generating a `uint32 <-> *uint32` conversion 3 lines above.

I think there may be a `mog` bug in basic type pointer-to/from-nonpointer field conversions that might fix this differently. :eyes:

Comment (hashi-derek) [reply proto/private/pbconfigentry/config_entry.gen.go:1837]: I had a chat with John about this on the side, and the `.proto` isn't using the `optional` keyword. Although, it sounded like the crashes still happened even after that flag was introduced.

File: agent/agent.go (hunk 2 above)
Comment (jmurret) [reply agent/agent.go:929]: good call.  added back inb3a057cb2e0da20cd37b7b49c1d1b1e5fb805f52

File: proto/private/pbconfigentry/config_entry.gen.go (hunk 3 above)
Comment (rboyer) [reply proto/private/pbconfigentry/config_entry.gen.go:1837]: If we land this mog thing the `*bool` behavior can come back: https://github.com/hashicorp/mog/pull/43

File: agent/agent.go [hunk 4]
```diff
@@ -921,29 +921,13 @@ func (a *Agent) Failed() <-chan struct{} {
...
-
 // configureXDSServer configures an XDS server with the proper implementation of
 // the PRoxyWatcher interface and registers the XDS server with Consul's
 // external facing GRPC server.
-func (a *Agent) configureXDSServer(proxyWatcher xds.ProxyWatcher) {
+func (a *Agent) configureXDSServer(proxyWatcher xds.ProxyWatcher, server *consul.Server) {
 	// TODO(agentless): rather than asserting the concrete type of delegate, we
 	// should add a method to the Delegate interface to build a ConfigSource.
-	if server, ok := a.delegate.(*consul.Server); ok {
+	if server != nil {
```
Comment (hashi-derek) [inline agent/agent.go:930]: Same here about reducing server pointers being passed

File: agent/agent.go [hunk 5]
```diff
@@ -4579,7 +4569,7 @@ func (a *Agent) listenerPortLocked(svcID structs.ServiceID, checkID structs.Chec
 	return port, nil
 }
 
-func (a *Agent) proxyDataSources() proxycfg.DataSources {
+func (a *Agent) proxyDataSources(server *consul.Server) proxycfg.DataSources {
```
[the remark this answers is not in the document]
Comment (rboyer) [reply agent/agent.go:4572]: That would be nice, but not like a mandatory thing, as we could chase that immediately after for x.x.1

**Review by rboyer — APPROVED** [review]

**Review by hashi-derek — APPROVED** [review]

**Pull request merged**